unison-sync/src/Unison/Sync/Types.hs

```diff
@@ -0,0 +1,515 @@
+import Data.Map (Map)
+import Data.Set (Set)
+import Data.Set.NonEmpty (NESet)
+import Data.Text (Text)
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+import Unison.Util.Tritraversable
+
+-- | A newtype for JSON encoding binary data.
+newtype Base64Bytes = Base64Bytes ByteString
+
+instance ToJSON Base64Bytes where
+  toJSON (Base64Bytes bytes) = String . Text.decodeUtf8 $ convertToBase Base64 bytes
+
+instance FromJSON Base64Bytes where
+  parseJSON = Aeson.withText "Base64" $ \txt -> do
+    either fail (pure . Base64Bytes) $ convertFromBase Base64 (Text.encodeUtf8 txt)
+
+newtype RepoName = RepoName Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+newtype HashJWT = HashJWT Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+data HashJWTClaims = HashJWTClaims
+  { hash :: Base32,
+    entityType :: EntityType
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON HashJWTClaims where
+  toJSON (HashJWTClaims hash entityType) =
+    object
+      [ "h" .= hash,
+        "t" .= entityType
+      ]
+
+instance FromJSON HashJWTClaims where
+  parseJSON = Aeson.withObject "HashJWTClaims" $ \obj -> do
+    hash <- obj .: "h"
+    entityType <- obj .: "t"
+    pure HashJWTClaims {..}
+
+newtype Base32 = Base32 Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON)
+
+newtype Hash = Hash Text
+  deriving newtype (Show, Eq, Ord, ToJSON, FromJSON, ToJSONKey, FromJSONKey)
+
+data TypedHash = TypedHash
+  { hash :: Hash,
+    entityType :: EntityType
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON TypedHash where
+  toJSON (TypedHash hash entityType) =
+    object
+      [ "hash" .= hash,
+        "type" .= entityType
+      ]
+
+instance FromJSON TypedHash where
+  parseJSON = Aeson.withObject "TypedHash" $ \obj -> do
+    hash <- obj .: "hash"
+    entityType <- obj .: "type"
+    pure $ TypedHash {..}
+
+data RepoPath = RepoPath
+  { repoName :: RepoName,
+    pathSegments :: [Text]
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON RepoPath where
+  toJSON (RepoPath name segments) =
+    object
+      [ "repo_name" .= name,
+        "path" .= Text.intercalate "." segments
+      ]
+
+instance FromJSON RepoPath where
+  parseJSON = Aeson.withObject "RepoPath" $ \obj -> do
+    repoName <- obj .: "repo_name"
+    pathSegments <- Text.splitOn "." <$> (obj .: "path")
+    pure RepoPath {..}
+
+newtype GetCausalHashByPathRequest = GetCausalHashByPathRequest
+  { repoPath :: RepoPath
+  }
+  deriving stock (Show, Eq, Ord)
+
+instance ToJSON GetCausalHashByPathRequest where
+  toJSON (GetCausalHashByPathRequest repoPath) =
+    object
+      [ "repo_path" .= repoPath
+      ]
+
+instance FromJSON GetCausalHashByPathRequest where
+  parseJSON = Aeson.withObject "GetCausalHashByPathRequest" $ \obj -> do
+    repoPath <- obj .: "repo_path"
+    pure GetCausalHashByPathRequest {..}
+
+newtype GetCausalHashByPathResponse = GetCausalHashByPathResponse
+  { causalHash :: HashJWT
+  }
```

Comment:
This should be a `Maybe` or else we need a different response for when a requested path doesn't exist.